keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -357,13 +393,15 @@ generated is valid OpenAPI v3.
 #### Beta
 
 - Native types are updated to capture capabilities introduced with v3
+  - anyOf is correctly used for IntOrString types
```

Comment:
looking at this:
```golang
// UnmarshalJSON implements the json.Unmarshaller interface.
// TODO: Remove support for leading/trailing whitespace
func (q *Quantity) UnmarshalJSON(value []byte) error {
	l := len(value)
	if l == 4 && bytes.Equal(value, []byte("null")) {
		q.d.Dec = nil
		q.i = int64Amount{}
		return nil
	}
	if l >= 2 && value[0] == '"' && value[l-1] == '"' {
		value = value[1 : l-1]
	}

	parsed, err := ParseQuantity(strings.TrimSpace(string(value)))
	if err != nil {
		return err
	}

	// This copy is safe because parsed will not be referred to again.
	*q = parsed
	return nil
}
```

Do I read this correctly that both a numeral like `1024` is allowed and also `"1k"`? That makes it another candidate we should add here.

Reply:
Some more to verify how they are represented today. All of them are polymorphic, and all from CRD types:
- `func (s *JSONSchemaPropsOrBool) UnmarshalJSON(data []byte) error`
- `func (s *JSONSchemaPropsOrStringArray) UnmarshalJSON(data []byte) error`
- `func (s *JSONSchemaPropsOrArray) UnmarshalJSON(data []byte) error`
- `func (s *JSON) UnmarshalJSON(data []byte) error`(for detault afaik)

Reply:
If I understand correctly, the marshaling/unmarshaling process works properly for the polymorphic types and it's just exposing the information in the OpenAPI? I reworded this to cover more than just int-or-string types.

Reply:
On a second look, it looks like we're generating the types in OpenAPI (currently marked as just "object") for these structs based on the go structs. I don't believe we're able to represent anyOf directly inside a go struct and we would need the help of annotations. I'm not sure if there's a better approach, but I think we may need something like 

XPropsOrBool, XPropsOrArray, etc similar to the XIntOrString annotation that we currently have.

Reply:
There are these
```
func (_ JSON) OpenAPISchemaType() []string {
	// TODO: return actual types when anyOf is supported
	return nil
}

// OpenAPISchemaFormat is used by the kube-openapi generator when constructing
// the OpenAPI spec of this type.
func (_ JSON) OpenAPISchemaFormat() string { return "" }
```

We need a more flexible variant that lets us return proper anyOf types.